libcaf_core/caf/log_event.hpp

```diff
@@ -0,0 +1,335 @@
+    else
+      return static_cast<uint64_t>(value);
+  }
+
+public:
+  // -- member types -----------------------------------------------------------
+
+  using list_type = detail::json::linked_list<log_event::field>;
+
+  using resource_type = detail::monotonic_buffer_resource;
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  explicit log_event_fields_builder(resource_type* resource) noexcept;
+
+  ~log_event_fields_builder() noexcept {
+    // nop
+  }
+
+  // -- add fields -------------------------------------------------------------
+
+  /// Adds a boolean or integer field.
+  template <class T>
+  std::enable_if_t<std::is_integral_v<T>>
+  add_field(std::string_view key, T value) {
+    auto& field = fields_.emplace_back(std::string_view{},
+                                       lift_integral(value));
+    field.key = deep_copy(key);
+  }
+
+  /// Adds a floating point field.
+  void add_field(std::string_view key, double value) {
+    auto& field = fields_.emplace_back(std::string_view{}, value);
+    field.key = deep_copy(key);
+  }
+
+  /// Adds a string field.
+  void add_field(std::string_view key, std::string_view value) {
+    auto& field = fields_.emplace_back(std::string_view{}, std::nullopt);
+    field.key = deep_copy(key);
+    field.value = deep_copy(value);
+  }
+
+  /// Adds a formatted string field.
+  template <class Arg, class... Args>
+  void add_field(std::string_view key, std::string_view fmt, Arg&& arg,
+                 Args&&... args) {
+    auto& field = fields_.emplace_back(std::string_view{}, std::nullopt);
+    field.key = deep_copy(key);
+    chunked_string_builder cs_builder{resource()};
+    chunked_string_builder_output_iterator out{&cs_builder};
+    detail::format_to(out, fmt, std::forward<Arg>(arg),
+                      std::forward<Args>(args)...);
+    field.value = cs_builder.build();
+  }
+
+  /// Adds nested fields.
+  template <class SubFieldsInitializer>
+  auto add_field(std::string_view key, SubFieldsInitializer&& init)
+    -> std::enable_if_t<std::is_same_v<
+      decltype(init(std::declval<log_event_fields_builder&>())), void>> {
+    auto& field = fields_.emplace_back(std::string_view{}, std::nullopt);
+    field.key = deep_copy(key);
+    log_event_fields_builder nested_builder{resource()};
+    init(nested_builder);
+    field.value = nested_builder.build();
+  }
+
+  // -- build ------------------------------------------------------------------
+
+  /// Seals the list and returns it.
+  [[nodiscard]] log_event::field_list build() {
+    return log_event::field_list{fields_.head()};
+  }
+
+private:
+  [[nodiscard]] resource_type* resource() noexcept {
+    return fields_.get_allocator().resource();
+  }
+
+  [[nodiscard]] std::string_view deep_copy(std::string_view str);
+
+  union { // Suppress the dtor call for fields_.
+    list_type fields_;
+  };
+};
+
+/// Builds a chunked string by allocating each chunk on a monotonic buffer.
```

Comment:
Copy-paste probably.
```suggestion
/// Builds a chunked log_event by allocating each field on a monotonic buffer.
```